File: src/main/java/org/nixos/idea/settings/NixLangSettingsConfigurable.java
package org.nixos.idea.settings;

import com.intellij.openapi.options.Configurable;
import com.intellij.openapi.options.ConfigurationException;
import com.intellij.openapi.options.SearchableConfigurable;
import com.intellij.openapi.util.NlsContexts;
import com.intellij.ui.RawCommandLineEditor;
import com.intellij.ui.TitledSeparator;
import com.intellij.ui.components.JBCheckBox;
import com.intellij.ui.components.JBLabel;
import com.intellij.util.ui.FormBuilder;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.nixos.idea.lsp.ui.CommandSuggestionsPopup;

import javax.swing.*;
import java.util.List;

public class NixLangSettingsConfigurable implements SearchableConfigurable, Configurable.Beta {
    private static final List<CommandSuggestionsPopup.Suggestion> BUILTIN_SUGGESTIONS = List.of(
            CommandSuggestionsPopup.Suggestion.builtin("<html>Use <b>nixpkgs-fmt</b> from nixpkgs</html>",
                    "nix --extra-experimental-features \"nix-command flakes\" run nixpkgs#nixpkgs-fmt")
    );

    private @Nullable JBCheckBox myEnabled;
    private @Nullable RawCommandLineEditor myCommand;
    private @Nullable JBLabel myTextArea;

    @Override
    public @NotNull @NonNls String getId() {
        return "org.nixos.idea.settings.NixLangSettingsConfigurable";
    }

    @Override
    public @NlsContexts.ConfigurableName String getDisplayName() {
        return "Nix";
    }

    @Override
    public @Nullable JComponent createComponent() {
        myEnabled = new JBCheckBox("Enable external formatter");
        myEnabled.addChangeListener(e -> updateUiState());

        myCommand = new RawCommandLineEditor();
        myCommand.getEditorField().getEmptyText().setText("Command to execute for formatting");
        myCommand.getEditorField().getAccessibleContext().setAccessibleName("Command to execute for formatting");

        myTextArea = new JBLabel();

        myTextArea.setText("Format Nix files via an external formatter. Source of focused file will be passed as standard input.");
        new CommandSuggestionsPopup(
                myCommand,
                NixLangSettings.getInstance().getCommandHistory(),
                BUILTIN_SUGGESTIONS
        ).install();


        return FormBuilder.createFormBuilder()
                .addComponent(new TitledSeparator("External Formatter Configuration"))
                .addComponent(myTextArea)
                .addComponent(myEnabled)
                .addLabeledComponent("Command: ", myCommand)
                .addComponentFillVertically(new JPanel(), 0)
                .getPanel();
    }

    @Override
    public void reset() {
        assert myEnabled != null;
        assert myCommand != null;

        NixLangSettings settings = NixLangSettings.getInstance();
        myEnabled.setSelected(settings.isFormatEnabled());
        myCommand.setText(settings.getFormatCommand());

        updateUiState();
    }

    @SuppressWarnings("UnstableApiUsage")
    @Override
    public void apply() throws ConfigurationException {
        assert myEnabled != null;
        assert myCommand != null;

        NixLangSettings settings = NixLangSettings.getInstance();
        settings.setFormatEnabled(myEnabled.isSelected());
        settings.setFormatCommand(myCommand.getText());
    }

    @Override
    public boolean isModified() {
        assert myEnabled != null;
        assert myCommand != null;

        NixLangSettings settings = NixLangSettings.getInstance();
        return Configurable.isCheckboxModified(myEnabled, settings.isFormatEnabled()) ||
                Configurable.isFieldModified(myCommand.getTextField(), settings.getFormatCommand());
    }

    private void updateUiState() {
        assert myEnabled != null;
        assert myCommand != null;

        myCommand.setEnabled(myEnabled.isSelected());
    }
}

File: src/main/java/org/nixos/idea/settings/ui/NixLangSettingsConfigurable.kt
package org.nixos.idea.settings.ui

import com.intellij.openapi.options.BoundSearchableConfigurable
import com.intellij.openapi.options.Configurable
import com.intellij.ui.RawCommandLineEditor
import com.intellij.ui.components.JBCheckBox
import com.intellij.ui.components.JBRadioButton
import com.intellij.ui.dsl.builder.AlignX
import com.intellij.ui.dsl.builder.Cell
import com.intellij.ui.dsl.builder.MAX_LINE_LENGTH_WORD_WRAP
import com.intellij.ui.dsl.builder.bind
import com.intellij.ui.dsl.builder.bindSelected
import com.intellij.ui.dsl.builder.panel
import com.intellij.ui.dsl.builder.selected
import com.intellij.ui.layout.and
import org.nixos.idea.settings.NixExternalFormatterSettings
import org.nixos.idea.settings.NixExternalFormatterSettings.InputMode
import org.nixos.idea.settings.NixExternalFormatterSettings.Strategy
import org.nixos.idea.settings.ui.UiDslExtensions.bindText
import org.nixos.idea.settings.ui.UiDslExtensions.placeholderText
import org.nixos.idea.settings.ui.UiDslExtensions.suggestionsPopup
import org.nixos.idea.settings.ui.UiDslExtensions.validateOnReset
import org.nixos.idea.settings.ui.UiDslExtensions.validateWhenTextChanged
import org.nixos.idea.settings.ui.UiDslExtensions.warnOnInput

class NixLangSettingsConfigurable :
    BoundSearchableConfigurable("Nix", "org.nixos.idea.settings.ui.NixLangSettingsConfigurable"),
    Configurable.Beta {

    override fun createPanel() = panel {
        group("External Formatter", indent = false) {
            val settings = NixExternalFormatterSettings.getInstance()
            lateinit var formatterEnabled: Cell<JBCheckBox>
            lateinit var customCommandEnabled: Cell<JBRadioButton>
            row {
                formatterEnabled = checkBox("Enable external formatter")
                    .bindSelected(settings::isEnabled)
                    .comment(
                        """
                        Format Nix files via an external formatter.
                        """.trimIndent(),
                        MAX_LINE_LENGTH_WORD_WRAP
                    )
            }
            indent {
                buttonsGroup {
                    row {
                        radioButton("<html>Use <b>nix fmt</b></html>", Strategy.NIX_FMT)
                        customCommandEnabled = radioButton("Use custom command", Strategy.CUSTOM_COMMAND)
                    }
                }.bind(settings::strategy)
                groupRowsRange(indent = false) {
                    row("Command:") {
                        cell(RawCommandLineEditor())
                            .bindText(settings::formatCommand)
                            .placeholderText("Command to execute for formatting")
                            .suggestionsPopup(settings.formatCommandHistory, BUILTIN_SUGGESTIONS)
                            .align(AlignX.FILL)
                            .validateOnReset()
                            .validateWhenTextChanged()
                            .warnOnInput("You have to specify the command of the formatter") {
                                it.text.isNullOrBlank()
                            }
                    }
                    buttonsGroup {
                        row {
                            radioButton("<html>Use <code>stdin</code> and <code>stdout</code></html>", InputMode.STDIO)
                                .comment("Source of the focused file will be passed as standard input.")
                        }
                        row {
                            radioButton("Use CLI argument", InputMode.CLI_ARGUMENT)
                                .comment("Path of the focused file will be specified as an additional argument.")
                        }
                    }.bind(settings::formatCommandInputMode)
                    row {
                        checkBox("Run formatter in the file's parent directory")
                            .bindSelected(settings::isFormatCommandRunInFileDirectory)
                            .comment("Required for <b>nix fmt</b> to work.")
                    }
                }.visibleIf(formatterEnabled.selected and customCommandEnabled.selected)
            }.enabledIf(formatterEnabled.selected)
        }
    }

    override fun apply() {
        super.apply()
        reset() // Update UI components to use normalized property values
    }
}

private val BUILTIN_SUGGESTIONS: List<CommandSuggestionsPopup.Suggestion> = listOf(
    CommandSuggestionsPopup.Suggestion.builtin(
        "<html>Use <b>nixpkgs-fmt</b> from nixpkgs</html>",
        "nix --extra-experimental-features \"nix-command flakes\" run nixpkgs#nixpkgs-fmt"
    ),
    CommandSuggestionsPopup.Suggestion.builtin(
        "<html>Use <b>nix fmt</b> from flake.nix</html>",
        "nix --extra-experimental-features \"nix-command flakes\" fmt"
    )
)
